convert2rhel/pkghandler.py

```diff
@@ -1101,52 +1068,173 @@ def clean_yum_metadata():
+        if re.findall(r".*(x86_64|s390x|i86|ppc64le|aarch64)$", release_arch):
+            release, arch = release_arch.rsplit(".", 1)
+        # package is in NEVR format
         else:
-            epoch, version = epoch_version
-        # Discard the arch
-        release = release_arch.rsplit(".", 1)[0]
-        pkg_ver_components.append((epoch, version, release))
+            release = release_arch
+            arch = None
+
+    # package is in either ENVR, ENVRA, NVR or NVRA
+    else:
+        # splitFilename doesn't work with packages in NEVRA/NEVR format, that's why we are using it only here.
+        (name, version, release, epoch, arch) = pkgmanager.splitFilename(pkg)
+
+        # splitFilename places part of the release in the arch field when no arch is present
+        # if arch field is invalid, append contents to release field
+        if arch not in ("x86_64", "s390x", "i86", "ppc64le", "aarch64"):
+            temp_release = arch
+            arch = None
+            release = "%s.%s" % (release, temp_release)
+
+    # convert any empty strings to None for consistency
+    pkg_ver_components = [(i or None) for i in (name, epoch, version, release, arch)]
+
+    return pkg_ver_components
+
+
+def _parse_pkg_with_dnf(pkg):
+    """
+    Parse verison string using hawkey and dnf.
+
+    :param pkg: The package to be parsed.
+    :type pkg: str
+    :return: Return a list containing name, epoch, version, release, arch
+    :rtype: list[str | None]
+    """
+
+    name = epoch = version = release = arch = None
+    no_arch_data = None
+
+    # if format is ENVR/ENVRA, store and remove epoch and evaluate as NVR/NVRA
+    if re.findall(r"^\d:", pkg):
+        pkg = pkg.split(":")
+        epoch = pkg[0]
+        pkg = pkg[1]
+
+    # returns generator for every possible nevra
+    subject = pkgmanager.dnf.subject.Subject(pkg)
+    possible_nevra = subject.get_nevra_possibilities(forms=[pkgmanager.hawkey.FORM_NEVRA, pkgmanager.hawkey.FORM_NEVR])
+
+    # loop through each possible set of nevra fields and select the valid one
+    for _, nevra in enumerate(possible_nevra):
+
+        # current arch is valid
+        if str(nevra.arch) in ("x86_64", "s390x", "i86", "ppc64le", "aarch64"):
+            name = nevra.name
+            epoch = epoch or nevra.epoch
+            version = nevra.version
+            release = nevra.release
+            arch = nevra.arch
+            break
+
+        # there is no arch present and current arch is None
+        if nevra.arch is None:
+            no_arch_data = nevra
+
+        # arch is not valid, move on to next iteration
+
+    else:  # This else goes with the for loop
+        name = no_arch_data.name
+        epoch = epoch or no_arch_data.epoch
+        version = no_arch_data.version
+        release = no_arch_data.release
+        arch = no_arch_data.arch
+
+    # convert epoch from integer to string for consistency
+    epoch = str(epoch)
+
+    if epoch == "None":
+        epoch = None
```

Comment:
I would rewrite this part here as: 

```suggestion
    if epoch is not None:
        epoch = str(epoch)
```

I know we went over this in our chat, but, I just tested the code, and it seems to me that `epoch` will be either an instance of `int` or `NoneType`. In that case, we could just verify that `epoch` is not None, and if it is not, cast it to string to keep it consistent. Otherwise, it is None already, just leave it as it is.

Reply:
This is much better thanks! I'll add this